compiler/next/include/chpl/Queries/Context.h

```diff
@@ -0,0 +1,352 @@
+
+Since the argument/key types are stored in a hashtable, we need
+``std::hash<KeyType>`` and ``std::equal_to<KeyType>`` to be implemented, e.g.
+
+.. code-block:: c++
+
+    namespace std {
+      template<> struct hash<chpl::MyPodKeyType> {
+        size_t operator()(const chpl::ast::UniqueString key) const {
+          return doSomethingToComputeHash...;
+        }
+      };
+      template<> struct equal_to<chpl::MyPodKeyType> {
+        bool operator()(const chpl::MyPodKeyType lhs,
+                        const chpl::MyPodKeyType rhs) const {
+          return doSomethingToCheckIfEqual...;
+        }
+      };
+    }
+
+The process of computing a query and checking to see if it maches a saved
+result requires that the result type implement ``chpl::combine``:
+
+.. code-block:: c++
+
+    namespace chpl {
+      template<> struct combine<MyResultType> {
+        bool operator()(chpl::ast::UniqueString& keep,
+                        chpl::ast::UniqueString& addin) const {
+          return doSomethingToCombine...;
+        }
+      };
+
+On entry to the ``combine`` function, ``keep`` is the current value in the
+program database and ``addin`` is the newly computed value. The ``combine``
+function needs to:
+
+  * store the current, updated result in ``keep``
+  * store the unused result in ``addin``
+  * return ``true`` if ``keep`` matched ``addin``; that is, ``keep`` did not
+    need to be updated.
+
+For most result types, ``return defaultCombine(keep, addin);`` should be
+sufficient. In the event that a result is actually a collection of results
+that *owns* the elements (for example, when parsing, the result is
+conceptually a vector of top-level symbol), the ``combine`` function
+should try to update only those elements of ``keep`` that changed by swapping
+in the appropriate elements from ``addin``. This strategy allows later queries
+that depend on such a result to use pointers to the owned elements and to
+avoid updating everything if just one element changed.
+
+Queries *can* return results that contain non-owning pointers to results from
+dependent queries. In that event, the combine function should not rely on the
+contents of these pointers. The system will make sure that they refer to valid
+memory but they might be a combination of old results. Additionally, the system
+will ensure that any old results being replaced will remain allocated until the
+garbage collection runs outside of any query.
+
+For example, a ``parse`` query might result in a list of ``owned`` AST element
+pointers. A follow-on ``listSymbols`` result in something containing these AST
+element pointers, but not owning them. The ``listSymbols`` query needs to use a
+``combine`` function that does not look into these queries.
+
+\endrst
+
+ */
+class Context {
+ private:
+  // map that supports uniqueCString / UniqueString
+   typedef std::unordered_map<const char*, char*, chpl::detail::UniqueStrHash, chpl::detail::UniqueStrEqual> UniqueStringsTableType;
```

Comment:
In C++11 and later, `using Type1 = Type2;` is generally preferred to `typedef Type2 Type1;`.